File: disjoint_set.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

class Node
{
public:
    int value;
    int rank;
    Node* parent;
    Node(int x = 0, int r = 0) : value(x), parent(this), rank(r) {}
};

Node* Find(Node* x)
{
    if(x->parent == x)
        return x->parent;
    else
    {
        x->parent = Find(x->parent);
        return x->parent;
    }
}

void Merge(Node* x, Node* y)
{
    if(x->rank < y->rank)
        x->parent = y;
    else
    {
        y->parent = x;
        if(x->rank == y->rank)
            x->rank += 1;
    }
}

void Union(Node* x, Node* y)
{
    if(x->parent != y->parent)
        Merge(x->parent, y->parent);
}

int main(int argc, char ** argv)
{
    unsigned int T;
    cin >> T;
    vector<vector<char> > M;
    vector<Node> L(T);

    for(int i = 0; i < T; i++)
    {
        L[i].value = i;
    }

    for(int i = 0; i < T; i++)
    {
        vector<char> tmp;
        for(int j = 0; j < T; j++)
        {
            char c;
            cin >> c;
            tmp.push_back(c);
        }
        M.push_back(tmp);
    }

    for(int i = 0; i < T; i++)
    {
        for(int j = 0; j < i; j++)
        {
            if(M[i][j] == 'Y')
                Union(&L[i], &L[j]);
        }
    }

    /*
    for(int i = 0; i < T; i++)
    {
        cout << L[i].parent << " ";
    }
    cout << endl;
    */

    set<Node*> S;
    for(int i = 0; i < T; i++)
    {
        S.insert(L[i].parent);
    }
    cout << S.size() << endl;
    return 0;
}

File: graph/disjoint_set.py
#!/usr/bin/env python

class Node:
    def __init__(self, x):
        self.value = x
        self.parent = self
        self.rank = 0
    def __str__(self):
        return str(self.value)

def Find(x):
    if x.parent == x:
        return x
    else:
        x.parent = Find(x.parent)
        return x.parent

def Union(x, y):
    if Find(x) != Find(y):
        Merge(Find(x), Find(y))

def Merge(x, y):
    if x.rank < y.rank:
        x.parent = y
    else:
        y.parent = x
        if x.rank == y.rank:
            x.rank = x.rank + 1

if __name__ == "__main__":
    T = input()
    L = [Node(x) for x in range(T)]
    M = []

    for i in xrange(T):
        line = raw_input().strip()
        cs = [c for c in line]
        M.append(cs)
    for i in xrange(T):
        for j in xrange(i):
            if M[i][j] == 'Y':
                Union(L[i], L[j])

    sets = set([str(Find(x)) for x in L])
    #print "sets = ", sets
    print len(sets)
